bug: can't compile caltech-other/voronoi under macOS

Hi,

I just stumbled upon a compilation error. I'm using **macOS BigSur 11.6.1 x64**. 

```shell
cd /Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4/caltech-other/voronoi
/usr/local/cm3/bin/cm3 -build -DBUILD_DIR=AMD64_DARWIN -DROOT=/Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4 -DTARGET=AMD64_DARWIN -DM3_BACKEND_MODE=C
--- building in AMD64_DARWIN ---

ignoring ../src/m3overrides

new source -> compiling Voronoi.i3
new source -> compiling Voronoi.m3
new source -> compiling edgelist.c
new source -> compiling geometry.c
new source -> compiling heap.c
new source -> compiling memory.c
new source -> compiling mod3_main.c
new source -> compiling output.c
../src/output.c: In function 'void clip_line(Edge*)':
../src/output.c:135:24: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]
  135 |                 if ((x1> pxmax & x2>pxmax) | (x1<pxmin&x2<pxmin)) return;
      |                      ~~^~~~~~~
../src/output.c:135:49: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]
  135 |                 if ((x1> pxmax & x2>pxmax) | (x1<pxmin&x2<pxmin)) return;
      |                                               ~~^~~~~~
../src/output.c:157:24: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]
  157 |                 if ((y1> pymax & y2>pymax) | (y1<pymin&y2<pymin)) return;
      |                      ~~^~~~~~~
../src/output.c:157:49: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]
  157 |                 if ((y1> pymax & y2>pymax) | (y1<pymin&y2<pymin)) return;
      |                                               ~~^~~~~~
cc1plus: all warnings being treated as errors
  compile_c => 1
```

The problem is obvious: missing parentheses when dealing with the **&-operator**.

--EDIT: the problem wasn't the missing spaces but instead the wrong operator. Should have been *&&* and not the bitwise variant.

Looks like it should be && and not & ?

On Tue, Jan 18, 2022 at 1:19 PM Harris Brakmić ***@***.***>
wrote:

> Hi,
>
> I just stumbled upon a compilation error. I'm using *macOS BigSur 11.6.1
> x64*.
>
> cd /Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4/caltech-other/voronoi
> /usr/local/cm3/bin/cm3 -build -DBUILD_DIR=AMD64_DARWIN -DROOT=/Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4 -DTARGET=AMD64_DARWIN -DM3_BACKEND_MODE=C
> --- building in AMD64_DARWIN ---
>
> ignoring ../src/m3overrides
>
> new source -> compiling Voronoi.i3
> new source -> compiling Voronoi.m3
> new source -> compiling edgelist.c
> new source -> compiling geometry.c
> new source -> compiling heap.c
> new source -> compiling memory.c
> new source -> compiling mod3_main.c
> new source -> compiling output.c
> ../src/output.c: In function 'void clip_line(Edge*)':../src/output.c:135:24: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]  135 |                 if ((x1> pxmax & x2>pxmax) | (x1<pxmin&x2<pxmin)) return;      |                      ~~^~~~~~~../src/output.c:135:49: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]  135 |                 if ((x1> pxmax & x2>pxmax) | (x1<pxmin&x2<pxmin)) return;      |                                               ~~^~~~~~../src/output.c:157:24: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]  157 |                 if ((y1> pymax & y2>pymax) | (y1<pymin&y2<pymin)) return;      |                      ~~^~~~~~~../src/output.c:157:49: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]  157 |                 if ((y1> pymax & y2>pymax) | (y1<pymin&y2<pymin)) return;      |                                               ~~^~~~~~cc1plus: all warnings being treated as errors  compile_c => 1
>
> The problem is obvious: missing parentheses when dealing with the
> *&-operator*.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/modula3/cm3/issues/897>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKYNJK43IOSRUBN33QRHYDUWXKOZANCNFSM5MIHC3LA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/caltech-other/voronoi/src/output.c
+++ b/caltech-other/voronoi/src/output.c
@@ -132,7 +132,7 @@ float x1,x2,y1,y2;
 			y2 = s2->coord.y;
 		if(y2<pymin) return;
 		x2 = e -> c - e -> b * y2;
-		if (((x1> pxmax) & (x2>pxmax)) | ((x1<pxmin)&(x2<pxmin))) return;
+		if ((x1> pxmax && x2>pxmax) | (x1<pxmin && x2<pxmin)) return;
 		if(x1> pxmax)
 		{	x1 = pxmax; y1 = (e -> c - x1)/e -> b;}
 		if(x1<pxmin)
@@ -154,7 +154,7 @@ float x1,x2,y1,y2;
 			x2 = s2->coord.x;
 		if(x2<pxmin) return;
 		y2 = e -> c - e -> a * x2;
-		if (((y1> pymax) & (y2>pymax)) | ((y1<pymin)&(y2<pymin))) return;
+		if (( y1>pymax && y2>pymax) | (y1<pymin && y2<pymin)) return;
 		if(y1> pymax)
 		{	y1 = pymax; x1 = (e -> c - y1)/e -> a;}
 		if(y1<pymin)

--- a/caltech-other/voronoi/src/output.c
+++ b/caltech-other/voronoi/src/output.c
@@ -154,7 +154,7 @@ float x1,x2,y1,y2;
 			x2 = s2->coord.x;
 		if(x2<pxmin) return;
 		y2 = e -> c - e -> a * x2;
-		if (( y1>pymax && y2>pymax) | (y1<pymin && y2<pymin)) return;
+		if ((y1>pymax && y2>pymax) | (y1<pymin && y2<pymin)) return;
 		if(y1> pymax)
 		{	y1 = pymax; x1 = (e -> c - y1)/e -> a;}
 		if(y1<pymin)

--- a/caltech-other/voronoi/src/output.c
+++ b/caltech-other/voronoi/src/output.c
@@ -132,7 +132,7 @@ float x1,x2,y1,y2;
 			y2 = s2->coord.y;
 		if(y2<pymin) return;
 		x2 = e -> c - e -> b * y2;
-		if ((x1> pxmax & x2>pxmax) | (x1<pxmin&x2<pxmin)) return;
+		if (((x1> pxmax) & (x2>pxmax)) | ((x1<pxmin)&(x2<pxmin))) return;
 		if(x1> pxmax)
 		{	x1 = pxmax; y1 = (e -> c - x1)/e -> b;}
 		if(x1<pxmin)
@@ -154,7 +154,7 @@ float x1,x2,y1,y2;
 			x2 = s2->coord.x;
 		if(x2<pxmin) return;
 		y2 = e -> c - e -> a * x2;
-		if ((y1> pymax & y2>pymax) | (y1<pymin&y2<pymin)) return;
+		if (((y1> pymax) & (y2>pymax)) | ((y1<pymin)&(y2<pymin))) return;
 		if(y1> pymax)
 		{	y1 = pymax; x1 = (e -> c - y1)/e -> a;}
 		if(y1<pymin)

